bin/build

```
@@ -0,0 +1,158 @@
+  set -x
+fi
+
+# shellcheck disable=SC1090,SC1091
+. "$(dirname "${BASH_SOURCE[0]}")/util.sh"
+
+layers_dir="$1"
+env_dir="$2/env"
+
+if compgen -G "${env_dir}/*" > /dev/null; then
+  for var in "${env_dir}"/*; do
+    declare "$(basename "${var}")=$(<"${var}")"
+  done
+fi
+
+apt_layer=$layers_dir/apt
+mkdir -p "$apt_layer"
+
+# Store which STACK we are running on in the cache to bust the cache if it changes
+if [ -f "$apt_layer"/STACK ]; then
+  CACHED_STACK=$(cat "$apt_layer/STACK")
+else
+  CACHED_STACK=$CNB_STACK_ID
+fi
+
+# Ensure we store the STACK in the cache for next time.
+echo "$CNB_STACK_ID" > "$apt_layer/STACK"
+
+APT_CACHE_DIR="$apt_layer/cache"
+APT_STATE_DIR="$apt_layer/state"
+APT_SOURCELIST_DIR="$apt_layer/sources"   # place custom sources.list here
+APT_SOURCES="$APT_SOURCELIST_DIR/sources.list"
+APT_VERSION=$(apt-get -v | awk 'NR == 1{ print $2 }')
+
+case "$APT_VERSION" in
+	0* | 1.0*) APT_FORCE_YES=(--force-yes);;
+	*)         APT_FORCE_YES=(--allow-downgrades --allow-remove-essential --allow-change-held-packages);;
+esac
+
+if [ -f "$apt_layer"/Aptfile ] && cmp -s Aptfile "$apt_layer"/Aptfile && [[ $CACHED_STACK == "$CNB_STACK_ID" ]] ; then
+	log::info "=> Aptfile and stack look the same, reusing cache"
+else
+	# Aptfile changed or does not exist or STACK changed
+	log::info "=> Detected Aptfile changes, flushing cache"
+	rm -rf "$APT_CACHE_DIR"
+	mkdir -p "$APT_CACHE_DIR/archives/partial"
+	mkdir -p "$APT_STATE_DIR/lists/partial"
+	mkdir -p "$APT_SOURCELIST_DIR"
+	cp -f Aptfile "$apt_layer"/Aptfile
+	cat "/etc/apt/sources.list" > "$APT_SOURCES"
+	if grep -q -e "^:repo:" Aptfile; then
+	log::info "=> Adding custom repositories"
+	grep -s -e "^:repo:" Aptfile | sed 's/^:repo:\(.*\)\s*$/\1/g' >> "$APT_SOURCES"
+	fi
+
+	APT_OPTIONS=(
+		-o debug::nolocking=true
+		-o dir::cache="$APT_CACHE_DIR"
+		-o dir::state="$APT_STATE_DIR"
+		-o dir::etc::sourcelist="$APT_SOURCES"
+		-o dir::etc::sourceparts=/dev/null
+	)
+
+	log::info "=> Updating apt caches"
+	# Ignore stderr from apt-get update containing: rm: cannot remove '/var/cache/apt/archives/partial/*.deb': Permission denied
+	apt-get "${APT_OPTIONS[@]}" update 2> >(grep -v "rm: cannot remove") | indent
+
+	while read -r line; do
+	  if [[ $line != "#"* ]] && [[ $line != ":repo:"* ]]; then
+		PACKAGE=$line
+		if [[ $PACKAGE == *deb ]]; then
+		  PACKAGE_NAME=$(basename "$PACKAGE" .deb)
+		  PACKAGE_FILE=$APT_CACHE_DIR/archives/$PACKAGE_NAME.deb
+
+		  log::info "=> Fetching $PACKAGE"
+		  curl --silent --show-error --fail -L -z "$PACKAGE_FILE" -o "$PACKAGE_FILE" "$PACKAGE" 2>&1 | indent
+		else
+		  log::info "=> Fetching .debs for $PACKAGE"
+		  apt-get "${APT_OPTIONS[@]}" -y "${APT_FORCE_YES[@]}" -d install --reinstall "$PACKAGE" | indent
+		fi
+	  fi
+	done < Aptfile
+
+	log::info "=> Installing apt packages with dpkg"
+	for DEB in "$APT_CACHE_DIR"/archives/*.deb
+	do
+		[[ -e "$DEB" ]] || break
+		basename "$DEB" | indent
+		dpkg -x "$DEB" "$apt_layer"/
```

```suggestion
		log::info "Installing $(basename "$DEB")"
		dpkg -x "$DEB" "$apt_layer"/
		echo
```

add a newline between each package for easier scrollback